source/brailleDisplayDrivers/freedomScientific.py

```diff
@@ -1,292 +1,495 @@
+
+	def _sendPacket(self, packetType, arg1=FS_BYTE_NULL, arg2=FS_BYTE_NULL, arg3=FS_BYTE_NULL, data=FS_DATA_EMPTY):
+		"""Send a packet to the display
+
+		@param packetType: Type of packet (first byte), use one of the FS_PKT constants
+		@type packetType: str
+		@param arg1: First argument (second byte of packet)
+		@type arg1: str
+		@param arg2: Second argument (third byte of packet)
+		@type arg2: str
+		@param arg3: Third argument (fourth byte of packet)
+		@type arg3: str
+		@param data: Data to send if this is an extended packet, required checksum will be added automatically
+		@type data: str
+		"""
+		def handleArg(arg):
+			if type(arg) == int:
+				return int2byte(arg)
+			return arg
+		arg1 = handleArg(arg1)
+		arg2 = handleArg(arg2)
+		arg3 = handleArg(arg3)
+		packet = [packetType, arg1, arg2, arg3, data]
+		if data:
+			packet.append(int2byte(self._calculateChecksum("".join(packet))))
+		self._dev.write("".join(packet))
+
+	def _onReceive(self, data):
+		if self.isUsb:
+			data = BytesIO(data)
+			packetType = data.read(1)
+		else:
+			packetType = data
+			data = self._dev
+
+		arg1 = data.read(1)
+		arg2 = data.read(1)
+		arg3 = data.read(1)
+		log.debug("Got packet of type %r with args: %r %r %r", packetType, arg1, arg2, arg3)
+		# Info and extended key responses are the only packets with payload and checksum
+		if packetType in (FS_PKT_INFO, FS_PKT_EXT_KEY):
+			length = ord(arg1)
+			payload = data.read(length)
+			checksum = ord(data.read(1))
+			calculatedChecksum = self._calculateChecksum(packetType + arg1 + arg2 + arg3 + payload)
+			assert calculatedChecksum == checksum, "Checksum mismatch, expected %s but got %s" % (checksum, payload[-1])
+		else:
+			payload = FS_DATA_EMPTY
+
+		self._handlePacket(packetType, arg1, arg2, arg3, payload)
+
+	def _handlePacket(self, packetType, arg1, arg2, arg3, payload):
+		"Handle a packet from the device"
+		if packetType == FS_PKT_ACK:
+			self._handleAck()
+		elif packetType == FS_PKT_NAK:
+			log.debugWarning("NAK received!")
+			self._handleAck()
+		elif packetType == FS_PKT_INFO:
+			self._manufacturer = payload[INFO_MANU_START:INFO_MANU_END].replace(FS_BYTE_NULL, "")
+			self._model = payload[INFO_MODEL_START:INFO_MODEL_END].replace(FS_BYTE_NULL, "")
+			self._firmwareVersion = payload[INFO_VERSION_START:INFO_VERSION_END].replace(FS_BYTE_NULL, "")
+			self.numCells = MODELS.get(self._model, 0)
+			if self.numCells in FOCUS_1_CELL_COUNTS:
+				# Focus 1: apply custom translation table
+				self.translationTable = FOCUS_1_TRANSLATION_TABLE
+			log.debug("Device info: manufacturer: %s model: %s, version: %s",
+				self._manufacturer, self._model, self._firmwareVersion)
+		elif packetType == FS_PKT_WHEEL:
+			# We use 0X7 to only get the last three bits
+			wheelNumber = ((ord(arg1) >> 3) & 0X7)
+			count = ord(arg1) & 0X7
+			if wheelNumber < 2:
+				isRight = False
+			else:
+				isRight = True
+			isDown = wheelNumber % 2 == 1
+			if isRight:
+				isDown = not isDown

```

Comment:
This block is quite confusing. How many wheels are expected? 

Checking this for values of wheelNumber shows the first 4 values as so:
```
0	 isRight: False	isDown: False
1	 isRight: False	isDown: True
2	 isRight: True	isDown: True
3	 isRight: True	isDown: False
```
The rest are the following pattern repeated, (`isRight` is never false again)
```
4	 isRight: True	isDown: True
5	 isRight: True	isDown: False
```
Is this correct?
If we only expect 4 values for wheel id I think a table would be prefereable:
```
isDown, isRight = [
	(False, False),
	(True, False),
	(True, True),
	(False, True)
][wheelID]
# remember to handle a raised indexError
```

Reply:
Funny that you mention this. This code was already in place when I started working at this driver, since the FS DLL also reported wheels in this way. So I never really thought if this was the best way to write this.

I think a list to lookup the wheel values makes this more readable, so I'll change this. There are only two wheels: left and right.

Reply:
> This code was already in place when I started working at this driver,

Yes, I understand it's not all code you wrote. However since we are here, and basically have to take the time to understand it anyway, we might as well try to make it clearer for the next person. 

> There are only two wheels: left and right.

This is also helpful information for someone who might not realise this assumption.